### DISCUSSION TITLE
Latency for http cache higher than expected

### DISCUSSION THREAD

**Discussion by anarazel** [cirruslabs/cirrus-ci-docs#986] [General]
Hi,

I was looking at using the http cache as a secondary ccache cache. Mainly with the goal of sharing cache state between multiple tasks - hard to do otherwise without each task overwriting cache state otherwise.  The results were a bit underwhelming so far - mostly due to the high latency when accessing the cache:

Log excerpt for a cache miss case:
[2022-03-20T21:17:12.178867 19169] No 549cob1s4q8tepm2scqn2sffpm904c3dk in http://127.0.0.1:12321/ccache2 (40.64 ms)
[2022-03-20T21:17:12.253862 19169] No 7108io3fin8bfnsbaee120hus3go26mc6 in http://127.0.0.1:12321/ccache2 (29.00 ms)
[2022-03-20T21:17:12.392228 19169] Stored 7108io3fin8bfnsbaee120hus3go26mc6 in http://127.0.0.1:12321/ccache2 (57.49 ms)
[2022-03-20T21:17:12.450723 19169] Stored 549cob1s4q8tepm2scqn2sffpm904c3dk in http://127.0.0.1:12321/ccache2 (57.45 ms)

This is with keepalive enabled, FWIW (which explains why the second miss is faster).

Is it that roughly the expected speed for http cache accesses?

For experimentation I enable ccache logging and uploaded it as an artifact...

empty cache run:
https://cirrus-ci.com/task/4791740338012160

full cache run:
https://cirrus-ci.com/task/5084119632707584

Greetings,

Andres

**Comment by anarazel** [comment]
Oh, one more question: What's the size of the http cache and when will it be cleaned up?

**Reply by fkorotkov** [reply]
Under the hood `cache` instruction is using the HTTP cache and therefore the retention period is the same and equal to 90 days. For managed by us tasks the size limit of a cache entry is 5Gb and there is no limit if you use your own infra.

Here is how the HTTP cache is working. Cirrus CI actually uses pre-signed URLs of services like GCS/S3/etc. So for a given request to download an cache entry Cirrus agent [requests a presigned URL](https://github.com/cirruslabs/cirrus-ci-agent/blob/b55ebe7c399dfd3ac68fb8d7eac5b031a9985f37/internal/http_cache/http_cache.go#L132) and then redirects to it. Everything on Cirrus side should be cached after the first request so that's why it was faster for your the second time. This design allows not to rely on Cirrus CI to be a proxy for all the caches and is beneficial to people with their own infra in different DC. 

I checked metrics for the `GenerateCacheDownloadURL` method on Cirrus side and p99 for it is around 18ms +-2ms. Not sure if we can go lower on the latency and have similarly scalable architecture. Do you have like a lot of small files you are trying to download/upload with high level of parallelism (there is also a limit of 4 simultaneous HTTP Cache requests per core)?

**Reply by fkorotkov** [reply]
On the second tough and after a bit deeper investigation I think we can improve the download latency with https://github.com/cirruslabs/cirrus-ci-agent/pull/221 But I don't think we can do anything about the upload latency.

**Reply by anarazel** [reply]
> Under the hood cache instruction is using the HTTP cache and therefore the retention period is the same and equal to 90 days. For managed by us tasks the size limit of a cache entry is 5Gb and there is no limit if you use your own infra.

Is there a way to *reduce* retention? For the type of caching we're talking about here there's not much point in retaining data for that long - it's extremely unlikely (if not impossible) to ever get a hit after a week or so.

I assume artifacts use the same infrastructure? If so, it'd be interesting to opt into a shorter retention there too.


> Do you have like a lot of small files you are trying to download/upload with high level of parallelism (there is also a limit of 4 simultaneous HTTP Cache requests per core)?

It's a parallel build (with #cores = 4). For each compiled file ccache tries to download one object and then another in case the first fails (different levels of caching). There's ~1.5k compiled files.

Right now we use ccache and download / upload the cache as part of the task using cache instructions. The problem is that that isn't shared across branches / PRs, and can't easily be made shared (because each run would otherwise overwrite the prior run as cirrus treats the whole cache as one "object" from what I can see). I hope to leverage ccache's HTTP cache support to have a "cross branch" cache, combined with a much smaller per-branch cache.


> On the second tough and after a bit deeper investigation I think we can improve the download latency with https://github.com/cirruslabs/cirrus-ci-agent/pull/221

Nice!

> But I don't think we can do anything about the upload latency.

The download latency is a much more significant issue for the ccache case, so that'd not be too bad.

**Reply by fkorotkov** [reply]
Could you please try your experiment once again? The change has been deployed.

**Reply by anarazel** [reply]
I reran a task and see some moderate gains. I can't see the agent version because the debug log appears to have been trimmed due to length - but the remaining output in the debug log differs, so I assume it's the new code.

Before:
[2022-03-20T21:20:38.716270 10726] Retrieved 4d06fk5oeet81d69vc9ivjvdsjd7vv8u0 from http://127.0.0.1:12321/ccache2 (72.65 ms)
Now:
[2022-03-23T17:25:01.304051 10618] Retrieved 6e684f3up57j58dmne3950vpt3rhibdca from http://127.0.0.1:12321/ccache2 (56.40 ms)

**Reply by anarazel** [reply]
> Is there a way to reduce retention? For the type of caching we're talking about here there's not much point in retaining data for that long - it's extremely unlikely (if not impossible) to ever get a hit after a week or so.

Looks like presigned URLs would actually make that easy enough? This isn't at all my area of expertise... If helpful I could open a feature request for configurable retention periods of caches?

**Reply by fkorotkov** [reply]
Yeah. Please create an feature request about custom cache retention and we'll see what we can do. There might some caveats with different storages.

Regarding the cache latency... There is one other thing we'll improve. It seems the agent is not reusing connections for fetching things via the presigned URLs. My bad! Fixing in https://github.com/cirruslabs/cirrus-ci-agent/pull/223

**Reply by anarazel** [reply]
> Yeah. Please create an feature request about custom cache retention and we'll see what we can do. There might some caveats with different storages.

#988

> Regarding the cache latency... There is one other thing we'll improve. It seems the agent is not reusing connections for fetching things via the presigned URLs. My bad! Fixing in https://github.com/cirruslabs/cirrus-ci-agent/pull/223

Heh. No worries.

Wonder if using something like ccache's secondary HTTP cache (and bazel and ...) end up eating too much per-request costs due to the much smaller requests...

**Reply by fkorotkov** [reply]
Ok. The latest agent was just deployed. Let's give it another try. I can't find any another bottlenecks simply by looking at the code. For further improvements we'll need to do proper profiling.

**Reply by anarazel** [reply]
I think something must have gone wrong - the times are way worse than before:
```
[2022-03-24T18:30:47.066501 10676] Retrieved 4d06fk5oeet81d69vc9ivjvdsjd7vv8u0 from http://127.0.0.1:12321/ccache2 (517.33 ms)
[2022-03-24T18:30:39.355728 10648] Retrieved 6e684f3up57j58dmne3950vpt3rhibdca from http://127.0.0.1:12321/ccache2 (1392.89 ms)
```
https://cirrus-ci.com/task/4611922271141888

Total time went from ~2min30s to 10min.

**Reply by fkorotkov** [reply]
Second run was 2:30 https://cirrus-ci.com/task/6132684358418432. I guess there was some weird issue with the first one but anyways there is not gains found.

**Reply by anarazel** [reply]
Phew ;). 

I think I'll give up on the plan to use the http cache for ccache - with cache entries mapping 1:1 to GCS/S3/etc I think the inherent latency (and likely cost) is just going to be too high. Before your [reply](https://github.com/cirruslabs/cirrus-ci-docs/discussions/986#discussioncomment-2417282), I hadn't realized that that's what the http cache maps to. I actually thought it'd just be served by a nearby instance and stored on instance storage or such (with in-memory caching).

**Reply by fkorotkov** [reply]
Maybe will be interesting to check if having a `micro` instance that is part of the free tier on GCP and have redis running there. It doesn't look like your overall ccache size is big ([seems just 50Mb](https://cirrus-ci.com/task/6132684358418432?logs=ccache_stat_post#L18)). Then you'll have an internal instance address that you can use from within your tasks.

**Reply by anarazel** [reply]
> Under the hood cache instruction is using the HTTP cache and therefore the retention period is the same and equal to 90 days. For managed by us tasks the size limit of a cache entry is 5Gb and there is no limit if you use your own infra.

Small clarification question: I assume that the 5GB are for caches, logs, explicit artifacts together? And that if the 5GB are exceeded the oldest cache/log/artifact is deleted?

Would be nice to document the 5GB on https://cirrus-ci.org/faq/#are-there-any-limits - I guess I'll open a ticket?


> Maybe will be interesting to check if having a micro instance that is part of the free tier on GCP and have redis running there.

Yea, I wondered about doing something like that. It'd be a bit of pain to deal with auth etc, given that one goal is to have random people be able to clone the repo and run CI tasks. But solvable.


> It doesn't look like your overall ccache size is big ([seems just 50Mb](https://cirrus-ci.com/task/6132684358418432?logs=ccache_stat_post#L18)). Then you'll have an internal instance address that you can use from within your tasks.

The cache is bigger than that normally, I just made it artificially smaller to test the http caching...

**Reply by JeremyRand** [reply]
> Small clarification question: I assume that the 5GB are for caches, logs, explicit artifacts together? And that if the 5GB are exceeded the oldest cache/log/artifact is deleted?

Also curious about this clarification.
